### DISCUSSION TITLE
1 MB firmware with asyncio and aiohttp

### DISCUSSION THREAD

**Discussion by LaySoft** [micropython/micropython#16022] [Core Development] [answered]
I have some relay devices that can only fit ESP-01. They currently run nodemcu firmware and communicate through websocket protocol with server. I would like to run MicroPython on these, but the 1MB flash size builds do not have the asyncio and aiohttp modules needed for websocket.

I tried to install them afterwards and aiohttp worked, but asyncio did not:
```
>>> import mip
>>> mip.install('asyncio')
Installing asyncio (latest) from https://micropython.org/pi/v2 to /lib
Package not found: https://micropython.org/pi/v2/package/6/asyncio/latest.json
Package may be partially installed
>>> 
```
How can I get asyncio to install? Is there any way I can build custom firmware myself with these two modules included and others missing?

**Comment by sophiedegran** [comment]
Have a look github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md

**Reply by LaySoft** [reply]
I read through the whole page, but I don't know what you mean that could help me with this particular problem...

**Comment by peterhinch** [comment]
Asyncio is built-in for those devices with enough resources to run it. Consequently it is not mip-installable. To install it on minimal hardware I think you will have to create a custom build.

**Reply by LaySoft** [reply]
I think it's not for lack of resources, but to save space, given that it says 1Mb in the name. Unfortunately there is no user-friendly builder for micropyhon like there is for [nodemcu](https://nodemcu-build.com/).

**Reply by Josverl** [reply]
There is mpbuild (in alfa) if using a cli works for you

**Reply by LaySoft** [reply]
Where can I find this build?

**Reply by scruss** [reply]
mpbuild is here: [mattytrentini/mpbuild: MicroPython build tool](https://github.com/mattytrentini/mpbuild) but it doesn't seem to support building any esp8266 boards.

[BUILD_CONTAINERS](https://github.com/mattytrentini/mpbuild/blob/235354e7bba9404da5a9f872621c8326fbc11fc1/src/mpbuild/build.py#L15) in build.py only mentions: esp32 mimxrt nrf renesas-ra rp2 samd stm32 unix

Some ESP-01s modules *may* have 4 MB flash chips on them: `esptool.py  flash_id` should identify them. MicroPython is a much larger (code size) project that NodeMCU, so I don't think you can do what you want to do with the hardware you have. Potential options may include:
* replacing the flash chips on your ESP-01 modules with 4 MB ones (difficult); or
* replacing your hardware with something like a [S2 mini](https://www.wemos.cc/en/latest/s2/s2_mini.html), a [Relay Shield](https://www.wemos.cc/en/latest/d1_mini_shield/relay.html#relay-shield) and a [Dual Base](https://www.wemos.cc/en/latest/d1_mini_shield/dual_base.html) (costs more; larger; soldering required)

**Reply by LaySoft** [reply]
- replacing the flash chips... You don't really mean that!

- replacing your hardware... This is not an option either, as I already have a [working fan controller](https://photos.app.goo.gl/dZCbVRiPyxTUyaGr6) with ESP-01 and a relay module that fits exactly, and I don't want to replace it.

Unfortunately it has stability problems, so I thought I'd try other software.

**Reply by jomasnash** [reply]
> I think it's not for lack of resources, but to save space, given that it says 1Mb in the name. Unfortunately there is no user-friendly builder for micropyhon like there is for [nodemcu](https://nodemcu-build.com/).

It's very easy to compile Micropython yourself using the Docker method (see: https://github.com/micropython/micropython/tree/master/ports/esp8266)

If you don't have Docker on your computer, use "Play With Docker (PWD)" (https://www.docker.com/play-with-docker/) , then you can use a fully installed Docker environment in your webbrowser,  4 hours for free (every time you login again you have 4 hours).

I can compile a 'fresh' Micropython in a few minutes on my computer/tablet/telephone. After compile, on your PWD, start a server:
"python -m http.server 80", then click on the PORT button of PWD and you can download your firmware.bin.

If you need additional files or scripts in PWD, you can use (win)scp to transfer these files from your computer to your PWD instance.

**Reply by scruss** [reply]
> * replacing the flash chips... You don't really mean that!

I absolutely do. I installed 3× 16 MB WinBond chips on some Black Pill boards yesterday. I'd swap out the chip from an ESP-01s but I only have 16 MB chips, and the largest that esptool thinks the board can take is 4 MB, so I'd be wasting most of a chip.

**Reply by robert-hh** [reply]
>  I'd swap out the chip from an ESP-01s

I did that once for an ESP-01 and a few times for other boards. The problem with the ESP01  was that I had the chip in the larger 208mil SOIC package instead of the 150mil package, so it was a little bit tricky. But there are flash chips available in the smaller package. Still hard for old eyes and shaky hands.

But it should be possible to make a build with asyncio and aiohttp for the 1MB flash. Only some other features would have to be dropped, like btree.

**Reply by robert-hh** [reply, marked as answer]
I made a build for a 1M device with asyncio and aiohttp, but without btree. Attached below. 
```
MicroPython v1.24.0-dirty on 2024-10-26; ESP module (1M) with ESP8266
Type "help()" for more information.
>>> help("modules")
__main__          binascii          io                re
_asyncio          builtins          json              requests/__init__
_boot             collections       lwip              select
_espnow           cryptolib         machine           socket
_onewire          deflate           math              ssl
_webrepl          dht               micropython       struct
aiohttp/__init__  ds18x20           mip/__init__      sys
aiohttp/aiohttp_ws                  errno             neopixel          time
apa102            esp               network           tls
array             espnow            ntptime           uasyncio
asyncio/__init__  flashbdev         onewire           uctypes
asyncio/core      framebuf          os                urequests
asyncio/event     gc                platform          vfs
asyncio/funcs     hashlib           port_diag         webrepl
asyncio/lock      heapq             pye               webrepl_setup
asyncio/stream    inisetup          random            websocket
Plus any modules on the filesystem
>>> 
```
[firmware.zip](https://github.com/user-attachments/files/17530788/firmware.zip)

**Reply by LaySoft** [reply]
Thank you very much! I tried it, it works fine, I'll test it for a few days and if everything is fine and stable, it will run on my fan controllers!
